hera_mc/cm_sysdef.py

```diff
@@ -6,123 +6,229 @@
+
+# Redefine dictionary as needed
+corr_index['parts_hera'] = 6
+corr_index['parts_paper'] = 10
+corr_index['parts_rfi'] = 4
+# polarizations should be one character
+all_pols['parts_hera'] = ['e', 'n']
+all_pols['parts_paper'] = ['e', 'n']
+all_pols['parts_rfi'] = ['e', 'n']
+redirect_part_types['parts_hera'] = ['node']
+single_pol_labeled_parts['parts_paper'] = ['cable-post-amp(in)', 'cable-post-amp(out)', 'cable-receiverator']
+
+# This generates the full_connection_path dictionary from port_def
+full_connection_path = {}
+for _x in port_def.keys():
+    ordered_path = {}
+    for k, v in six.iteritems(port_def[_x]):
+        ordered_path[v['position']] = k
+    sorted_keys = sorted(list(ordered_path.keys()))
+    full_connection_path[_x] = []
+    for k in sorted_keys:
+        full_connection_path[_x].append(ordered_path[k])
+
+
+def handle_redirect_part_types(part, at_date='now', session=None):
+    """
+    This handles the "special cases" by feeding a new part list back to hookup.
+    """
+    hpn_list = []
+    if part.part_type.lower() == 'node':
+        from hera_mc import cm_handling, cm_utils
+        rptc = cm_handling.Handling(session)
+        conn = rptc.get_part_connection_dossier(part.hpn, part.rev, port='all', at_date=at_date, exact_match=True)
+        redirect_list = []
+        for _k in conn.keys():
+            _pk = cm_utils.split_part_key(_k)[0]
+            if _pk.upper() == part.hpn.upper():
+                redirect_list.append(_k)
+        for _r in redirect_list:
+            for _x in conn[_r].keys_up:
+                if _x.upper().startswith('SNP'):
+                    hpn_list.append(cm_utils.split_connection_key(_x)[0])
+    return hpn_list
+
+
+def find_hookup_type(part_type, hookup_type):
+    if hookup_type in port_def.keys():
+        return hookup_type
+    if hookup_type is None:
+        for hookup_type in checking_order:
+            if part_type in port_def[hookup_type].keys():
+                return hookup_type
+    raise ValueError("hookup_type {} is not found.".format(hookup_type))
+
+
+def setup(part, port_query='all', hookup_type=None):
     """
     Given the current part and port_query (which is either 'all', 'e', or 'n')
-    this figures out which pols to do.  Basically, given 'pol' and part it
-    figures out whether to return ['e'], ['n'], ['e', 'n'], etc
+    this figures out which pols to do.  Basically, given the part and query it
+    figures out whether to return ['e*'], ['n*'], or ['e*', 'n*']
 
     Parameter:
     -----------
     part:  current part dossier
     port_query:  the ports that were requested ('e' or 'n' or 'all')
+    hookup_type:  if not None, will use specified hookup_type
+                  otherwise it will look through in order
     """
-
-    all_pols_lo = [x.lower() for x in all_pols]
+    if hookup_type is None:
+        hookup_type = find_hookup_type(part.part_type, None)
+    global this_hookup_type, all_pols, pind
```

Comment:
globals worry me a little, why are they required here?

Reply:
I agree - this should probably become a class.   (I may do one more clean-up PR for this).  This gets called multiple times in a hookup pass and the globals are specified at the start.  It might well be that I don't need those declarations, but in the absence of surety I opted to include it.

Reply:
Ok. I think it would be better to make it an object, what I'm most worried about is knock-on effects where it hangs around and affects later calls, but if you've tested that I can live with it short term (maybe make an issue to convert it to an object?).